MAINT: ensure towncrier can be run >1x, and is included in `spin docs`

rgommers opened the pull request:
This allows accumulating snippets into the main release notes file and committing the result during a release cycle. It will also make Sphinx formatting errors, cross-link issues, etc. visible when docs get built locally. This will allow fixing them as we go, rather than leaving it broken in devdocs and then having to fix it in one go once `towncrier` is run once right before or after creating a release branch.

xref gh-25921 and gh-25827 for recent issues with this.

rgommers committed:
MAINT: ensure towncrier can be run >1x, and is included in `spin docs`

This allows accumulating snippets into the main release notes file and
committing the result during a release cycle. It will also make Sphinx
formatting errors, cross-link issues, etc. visible when docs get built
locally. This will allow fixing them as we go, rather than leaving it
broken in devdocs and then having to fix it in one go once `towncrier`
is run once right before or after creating a release branch.

xref gh-25921 and gh-25827 for recent issues with this.

[skip azp] [skip actions] [skip cirrus]

rgommers commented:
CI doc build looks happy too: https://output.circle-artifacts.com/output/job/9f42cbe9-a4a5-43bc-a690-eef35559e762/artifacts/0/doc/build/html/release/2.0.0-notes.html

That will fix the now-missing notes at https://numpy.org/devdocs/release/2.0.0-notes.html.

charris commented:
I am not clear on how `notes-towncrier.rst` is supposed to be used.

mattip commented:
I think this is as good as it gets while using towncrier. There is some cruft left at [the bottom of the release notes](https://output.circle-artifacts.com/output/job/9f42cbe9-a4a5-43bc-a690-eef35559e762/artifacts/0/doc/build/html/release/2.0.0-notes.html#any-and-all-return-booleans-for-object-arrays) but after branching the notes are tweaked anyway.

seberg commented on doc/release/upcoming_changes/template.rst:
```diff
@@ -1,7 +1,4 @@
-{{ "=" * title|length }}
-{{ title }}
-{{ "=" * title|length }}
+**Content from release note snippets in doc/release/upcoming_changes:**
```
The setup makes sense to me, but besides it being done for 2.0, running things multiple time is just painful with or without this, since the sections will get messy.

So I think this line should be in the release notes file, the title I am happy to remove.  That way, when 2.0 is branched we create  new file with:

```
NumPy 2.1 release notes
=======================

.. inlcude: filename
```
and add this line if someone actually generates it for some reason intermittendly.

rgommers replied:
Updated, and fixed the template for new release notes too.

charris commented:
Note that the release notes also require `.. currentmodule:: numpy` on the first line when the release happens.

rgommers commented:
> Note that the release notes also require `.. currentmodule:: numpy` on the first line when the release happens.

Fixed that in `doc/source/release/template.rst` so it cannot be forgotten about anymore. 

I also updated the release walkthrough document.

mattip commented on doc/RELEASE_WALKTHROUGH.rst:
```diff
@@ -94,10 +94,12 @@ needed information.
+``doc/source/release/notes-towncrier.rst`` file into the release notes file
+(e.g., ``doc/source/release/2.3.4-notes.rst``), and delete the now-processed
+snippets in ``doc/release/upcoming_changes/``. This is safe to do multiple
+times during a release cycle.
```
I can't suggest changes below line 105, so I am commenting here: do the last lines of this paragraph need updating?

mattip reviewed (approved):
LGTM, one small question about the RELEASE_WALKTHROUGH.

rgommers committed:
DOC: remove no longer needed steps from release walkthrough

[ci skip]

seberg commented on doc/source/release/template.rst:
```diff
@@ -1,45 +1,21 @@
-
-New Features
-============
+*We'll choose highlights for this release near the end of the release cycle.*
```
We actually have a highlights section also in the release notes fragments which I think makes sense to not forget a highlight.  OTOH, I am not sure we ever actually used it so this is fine also.

seberg replied:
Sure, you certainly have to audit it at the end.

seberg reviewed (approved):
Thanks, nice to just always build it and it doesn't seem to make the release process harder.
I still don't like the idea of committing partial changes except maybe very late in the release cycle, but it doesn't matter.

In either case @charris is the one who dealt with this the most in the past and currently, so I think he should sign off/merge.

charris commented:
Well, let's give it a shot. Thanks Ralf.
